{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import scanpy as sc\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "from anndata import AnnData\n",
    "import pathlib\n",
    "import matplotlib.pyplot as plt\n",
    "import matplotlib as mpl\n",
    "import seaborn as sns\n",
    "import tangram as tg\n",
    "import os"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "data_root = \"./data/embyo_E8.5\"\n",
    "\n",
    "seq_data = sc.read_h5ad(f\"{data_root}_sc.h5ad\")\n",
    "spatial_data = sc.read_h5ad(f\"{data_root}_st.h5ad\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "spatial_data.var_names = [x.lower() for x in spatial_data.var_names]\n",
    "seq_data.var_names = [x.lower() for x in seq_data.var_names]\n",
    "\n",
    "spatial_data.var_names_make_unique()\n",
    "seq_data.var_names_make_unique()\n",
    "\n",
    "sc.pp.calculate_qc_metrics(seq_data, percent_top=None, log1p=False, inplace=True)\n",
    "mt_genes = seq_data.var_names.str.startswith(\"mt-\")\n",
    "seq_data = seq_data[seq_data.obs[\"total_counts\"] <= 33000, ~mt_genes].copy()\n",
    "sc.pp.filter_genes(seq_data, min_cells=int(0.01*seq_data.shape[0]))\n",
    "seq_data.X = seq_data.X.toarray()\n",
    "\n",
    "sc.pp.calculate_qc_metrics(spatial_data, percent_top=None, log1p=False, inplace=True)\n",
    "spatial_data = spatial_data[spatial_data.obs[\"total_counts\"] <= 600, :].copy()\n",
    "\n",
    "\n",
    "adata = spatial_data.copy()\n",
    "RNAseq_adata = seq_data.copy()\n",
    "\n",
    "# preprocess RNAseq data\n",
    "sc.pp.normalize_total(RNAseq_adata)\n",
    "sc.pp.log1p(RNAseq_adata)\n",
    "\n",
    "# subset spatial data into shared genes\n",
    "gene_names = np.intersect1d(adata.var_names, RNAseq_adata.var_names)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "adata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "RNAseq_adata"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We prepares the data using `pp_adatas`, which does the following:\n",
    "- Takes a list of genes from user via the `genes` argument. These genes are used as training genes.\n",
    "- Annotates training genes under the `training_genes` field, in `uns` dictionary, of each AnnData. \n",
    "- Ensure consistent gene order in the datasets (_Tangram_ requires that the the $j$-th column in each matrix correspond to the same gene).\n",
    "- If the counts for a gene are all zeros in one of the datasets, the gene is removed from the training genes.\n",
    "- If a gene is not present in both datasets, the gene is removed from the training genes.\n",
    "- In the pp_adatas function, the gene names are converted to lower case to get rid of the inconsistent capitalization. If this is not wanted, you can set the parameter gene_to_lowercase = False "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "tg.pp_adatas(RNAseq_adata, adata, genes=gene_names)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def cv_data_gen(adata_sc, adata_sp, cv_mode=\"loo\"):\n",
    "    \"\"\" Generates pair of training/test gene indexes cross validation datasets\n",
    "\n",
    "    Args:\n",
    "        adata_sc (AnnData): single cell data\n",
    "        adata_sp (AnnData): gene spatial data\n",
    "        mode (str): Optional. support 'loo' and '10fold'. Default is 'loo'.\n",
    "\n",
    "    Yields:\n",
    "        tuple: list of train_genes, list of test_genes\n",
    "    \"\"\"\n",
    "\n",
    "    # Check if training_genes key exist/is valid in adatas.uns\n",
    "    if \"training_genes\" not in adata_sc.uns.keys():\n",
    "        raise ValueError(\"Missing tangram parameters. Run `pp_adatas()`.\")\n",
    "\n",
    "    if \"training_genes\" not in adata_sp.uns.keys():\n",
    "        raise ValueError(\"Missing tangram parameters. Run `pp_adatas()`.\")\n",
    "\n",
    "    if not list(adata_sp.uns[\"training_genes\"]) == list(adata_sc.uns[\"training_genes\"]):\n",
    "        raise ValueError(\n",
    "            \"Unmatched training_genes field in two Anndatas. Run `pp_adatas()`.\"\n",
    "        )\n",
    "\n",
    "    genes_array = np.array(adata_sp.uns[\"training_genes\"])\n",
    "\n",
    "    if cv_mode == \"loo\":\n",
    "        cv = LeaveOneOut()\n",
    "    elif cv_mode == \"5fold\":\n",
    "        cv = KFold(n_splits=5)\n",
    "\n",
    "    for train_idx, test_idx in cv.split(genes_array):\n",
    "        train_genes = list(genes_array[train_idx])\n",
    "        test_genes = list(genes_array[test_idx])\n",
    "        yield train_genes, test_genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from tqdm import tqdm\n",
    "from sklearn.model_selection import LeaveOneOut\n",
    "from sklearn.model_selection import KFold\n",
    "from tangram import compare_spatial_geneexp\n",
    "\n",
    "cv_mode = \"5fold\"\n",
    "mode=\"cells\"\n",
    "#mode=\"clusters\"\n",
    "\n",
    "if mode==\"clusters\":\n",
    "    #whether or not cluter scRNA data\n",
    "    RNAseq_adata_label = RNAseq_adata.copy()\n",
    "    sc.pp.highly_variable_genes(RNAseq_adata_label)\n",
    "    RNAseq_adata_label = RNAseq_adata[:, RNAseq_adata_label.var.highly_variable].copy()\n",
    "    sc.pp.scale(RNAseq_adata_label, max_value=10)\n",
    "    sc.tl.pca(RNAseq_adata_label)\n",
    "    sc.pp.neighbors(RNAseq_adata_label)\n",
    "    sc.tl.leiden(RNAseq_adata_label, resolution = 0.5)\n",
    "    RNAseq_adata.obs['leiden'] = RNAseq_adata_label.obs.leiden\n",
    "    del RNAseq_adata_label\n",
    "    RNAseq_adata = tg.adata_to_cluster_expression(RNAseq_adata, \"leiden\", scale=True, add_density=False)\n",
    "\n",
    "verbose=False\n",
    "test_genes_list = []\n",
    "test_pred_list = []\n",
    "test_score_list = []\n",
    "train_score_list = []\n",
    "test_df_list = []\n",
    "curr_cv_set = 1\n",
    "\n",
    "if cv_mode == \"loo\":\n",
    "    length = len(list(RNAseq_adata.uns[\"training_genes\"]))\n",
    "elif cv_mode == \"5fold\":\n",
    "    length = 5\n",
    "\n",
    "for train_genes, test_genes in tqdm(\n",
    "    cv_data_gen(RNAseq_adata, adata, cv_mode), total=length\n",
    "):\n",
    "    # train\n",
    "    adata_map = tg.map_cells_to_space(RNAseq_adata,adata,\n",
    "                          cv_train_genes=train_genes,\n",
    "                          mode=mode,\n",
    "                          #cluster_label=\"leiden\",\n",
    "                          num_epochs=2000,\n",
    "                          learning_rate=0.1,\n",
    "                          verbose=False,\n",
    "                          device=\"cuda:1\"\n",
    "                          #device='cpu',\n",
    "                        )\n",
    "    \n",
    "    cv_genes = train_genes + test_genes\n",
    "\n",
    "    # project on space\n",
    "    adata_ge = tg.project_genes(\n",
    "        adata_map, RNAseq_adata[:, cv_genes], scale=False,\n",
    "    )\n",
    "\n",
    "    # retrieve result for test gene (genes X cluster/cell)\n",
    "    #if cv_mode == \"loo\":\n",
    "    adata_ge_test = adata_ge[:, test_genes].X.T\n",
    "    test_pred_list.append(adata_ge_test)\n",
    "\n",
    "    # output test genes dataframe\n",
    "\n",
    "    df_g = compare_spatial_geneexp(adata_ge, adata, RNAseq_adata, cv_genes)\n",
    "\n",
    "    test_df = df_g[df_g.index.isin(test_genes)]\n",
    "    test_score = df_g.loc[test_genes][\"score\"] \n",
    "\n",
    "    # output avg score\n",
    "    test_genes_list.append(test_genes)\n",
    "    test_score_list.append(test_score)\n",
    "    #train_score_list.append(train_score)\n",
    "    test_df_list.append(test_df)\n",
    "\n",
    "    if verbose == True:\n",
    "        msg = \"cv set: {}----test score: {:.3f}\".format(\n",
    "            curr_cv_set,  test_score\n",
    "        )\n",
    "        print(msg)\n",
    "\n",
    "    curr_cv_set += 1\n",
    "\n",
    "# use nanmean to ignore nan in score list\n",
    "if cv_mode == \"loo\":\n",
    "    avg_test_score = np.nanmean(test_score_list)\n",
    "else:\n",
    "    test_score_df = pd.concat(test_score_list, axis=0)\n",
    "    avg_test_score = np.nanmean(test_score_df)\n",
    "#avg_train_score = np.nanmean(train_score_list)\n",
    "\n",
    "cv_dict = {\n",
    "    \"avg_test_score\": avg_test_score\n",
    "}\n",
    "\n",
    "print(\"cv avg test score {:.3f}\".format(avg_test_score))\n",
    "#print(\"cv avg train score {:.3f}\".format(avg_train_score))\n",
    "\n",
    "if cv_mode == \"loo\":\n",
    "\n",
    "    # output df_test_genes dataframe\n",
    "    test_gene_df = pd.concat(test_df_list, axis=0)\n",
    "    \n",
    "    # output AnnData for generated spatial data by LOOCV\n",
    "    adata_ge_cv = sc.AnnData(\n",
    "        X=np.squeeze(test_pred_list).T,\n",
    "        obs=adata.obs.copy(),\n",
    "        var=pd.DataFrame(\n",
    "            test_score_list,\n",
    "            columns=[\"test_score\"],\n",
    "            index=np.squeeze(test_genes_list),\n",
    "        ),\n",
    "    )\n",
    "else:\n",
    "    # output df_test_genes dataframe\n",
    "    test_gene_df = pd.concat(test_df_list, axis=0)\n",
    "    \n",
    "    # output AnnData for generated spatial data by LOOCV\n",
    "    adata_ge_cv = sc.AnnData(\n",
    "        X=np.vstack(test_pred_list).T,\n",
    "        obs=adata.obs.copy(),\n",
    "        var=pd.DataFrame(\n",
    "            test_score_df.values,\n",
    "            columns=[\"test_score\"],\n",
    "            index=test_score_df.index,\n",
    "        ),\n",
    "    )\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "adata_ge_cv"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "adata_ge_cv.var['test_score'].values"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "cv_genes = adata_ge_cv.var[\"test_score\"].index.values\n",
    "adata_ge_cv.uns[\"train_genes_df\"] = pd.DataFrame(adata_ge_cv.var['test_score'].values,cv_genes, columns=[\"train_score\"])\n",
    "adata_ge_cv.uns[\"train_genes_df\"][\"sparsity_sc\"] = RNAseq_adata[\n",
    "        :, cv_genes\n",
    "    ].var.sparsity\n",
    "adata_ge_cv.uns[\"train_genes_df\"][\"sparsity_sp\"] = adata[\n",
    "        :, cv_genes\n",
    "    ].var.sparsity\n",
    "adata_ge_cv.uns[\"train_genes_df\"][\"sparsity_diff\"] = (\n",
    "        adata[:, cv_genes].var.sparsity\n",
    "        - RNAseq_adata[:, cv_genes].var.sparsity\n",
    "    )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from sklearn.metrics import mean_squared_error\n",
    "from scipy import stats\n",
    "\n",
    "pearson = []\n",
    "spearman = []\n",
    "kendalltau = []\n",
    "RMSE = []\n",
    "    \n",
    "\n",
    "for gene in gene_names:\n",
    "    v1 = adata[:,gene].X\n",
    "    v2 = adata_ge_cv[:,gene].X\n",
    "    personR = stats.pearsonr(v1.reshape(-1), v2.reshape(-1))\n",
    "    spearmanR = stats.spearmanr(v1.reshape(-1), v2.reshape(-1))\n",
    "    kentou = stats.kendalltau(v1.reshape(-1), v2.reshape(-1))\n",
    "    rmse = mean_squared_error(v1, v2, squared=False)\n",
    "\n",
    "    pearson.append(personR[0])\n",
    "    spearman.append(spearmanR[0])\n",
    "    kendalltau.append(kentou.statistic)\n",
    "    RMSE.append(rmse)\n",
    "\n",
    "norm_raw = stats.zscore(adata.X, axis=0)\n",
    "norm_imputed = stats.zscore(adata_ge_cv.X, axis=0)\n",
    "\n",
    "norm_rmse = []\n",
    "for v1, v2 in zip(norm_raw.T, norm_imputed.T):\n",
    "    rmse = mean_squared_error(v1, v2, squared=False)\n",
    "    norm_rmse.append(rmse)\n",
    "\n",
    "df_sc = pd.DataFrame({\"Pearson\": pearson, \"Spearman\": spearman, \"Kendalltou\":kendalltau, \"norm_RMSE\": norm_rmse,\"RMSE\":RMSE})\n",
    "df_sc.mean()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "adata_ge_cv.var[\"Pearson\"] = pearson\n",
    "adata_ge_cv.var[\"Spearman\"] = spearman\n",
    "adata_ge_cv.var[\"Kendall_tau\"] = kendalltau\n",
    "adata_ge_cv.var[\"norm_RMSE\"] = norm_rmse\n",
    "adata_ge_cv.var[\"RMSE\"] = RMSE"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "interpreter": {
   "hash": "d6a69e99b54b137c0a46b7eb1c01877b00eaeac81f3c7d96786977ff3fe6ebe1"
  },
  "kernelspec": {
   "display_name": "tissue",
   "language": "python",
   "name": "tissue"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.18"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
